File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/forbid_greater_than_comparison.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Forbid greater than comparison rule."""

from pathlib import Path

import libcst as cst

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import parse_python, relative, severity_for, stack_sources


class ForbidGreaterThanComparisonRule(HarnessCheckRule):
    """Validate forbidGreaterThanComparison check."""

    category = "forbidGreaterThanComparison"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> list[Finding]:
        """Validate forbidGreaterThanComparison check."""
        severity = severity_for(manifest, self.category)
        sources = stack_sources(root, manifest, self.category)
        class _ComparisonFinder(cst.CSTVisitor):
            def __init__(self) -> None:
                super().__init__()
                self.findings: list[Finding] = []
            def visit_Comparison(self, node: cst.Comparison) -> bool:
                for target in node.comparators:
                    if isinstance(target.operator, (cst.GreaterThan, cst.GreaterThanEqual)):
                        pos = self.get_metadata(cst.metadata.PositionProvider, node)
                        self.findings.append(Finding(
                            severity,
                            self.category,
                            f"{relative(path)}:{pos.start.line}: forbidden `>`/`>=` comparison; rewrite with `<`/`<=` so the smaller value is on the left",
                        ))
                return True
        result = []
        for path in sources:
            tree, error = parse_python(path)
            if error is not None:
                result.append(Finding(
                    severity,
                    self.category,
                    f"{relative(path)}: syntax error: {error}",
                ))
                continue
            wrapper = cst.MetadataWrapper(tree)
            visitor = _ComparisonFinder()
            wrapper.visit(visitor)
            result.extend(visitor.findings)
        return result

File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/forbid-greater-than-comparison.ts
#!/usr/bin/env bun
import {
  SyntaxKind,
  createSourceFile,
  forEachChild,
  isBinaryExpression,
  type Node,
  type SourceFile,
} from "typescript@6.0.3";
import type { Finding, HarnessCheckRule, HarnessManifest, RuleContext } from "../harness-check-rule";

/**
 * Forbid greater-than comparisons in TypeScript.
 */
export class ForbidGreaterThanComparisonRule implements HarnessCheckRule {
  static readonly category = "forbidGreaterThanComparison";

  constructor(private readonly ctx: RuleContext) {}

  applies(_manifest: HarnessManifest): boolean {
    return true;
  }

  validate(_root: string, manifest: HarnessManifest): Finding[] {
    const sources = this.ctx.stackSources(manifest, "typescript");
    return sources.flatMap((file) => {
      const text = this.ctx.read(file);
      if (!text) {
        return [];
      }
      let sourceFile: SourceFile;
      try {
        sourceFile = createSourceFile(file, text, SyntaxKind.LatestVersion, true);
      } catch {
        return [
          {
            severity: this.ctx.severityOf(manifest, ForbidGreaterThanComparisonRule.category),
            category: ForbidGreaterThanComparisonRule.category,
            message: `failed to parse TypeScript: ${file}`,
          },
        ];
      }
      const findings: Finding[] = [];
      const visit = (node: Node): void => {
        if (isBinaryExpression(node)) {
          const kind = node.operatorToken.kind;
          if (kind === SyntaxKind.GreaterThanToken || kind === SyntaxKind.GreaterThanEqualsToken) {
            const { line } = sourceFile.getLineAndCharacterOfPosition(node.operatorToken.getStart(sourceFile));
            const operator = kind === SyntaxKind.GreaterThanToken ? ">" : ">=";
            findings.push({
              severity: this.ctx.severityOf(manifest, ForbidGreaterThanComparisonRule.category),
              category: ForbidGreaterThanComparisonRule.category,
              message: `${file}:${line + 1}: forbidden \`${operator}\`; use \`${operator === ">" ? "<" : "<="}\``,
            });
          }
        }
        forEachChild(node, visit);
      };
      visit(sourceFile);
      return findings;
    });
  }
}
